### PR TITLE
Fix findfirst with decreasing step range

### LINKED ISSUE #33808: Buggy findfirst(p::Union{Fix2{typeof(==),T}, Fix2{typeof(isequal),T}}, r::StepRange{T,S}) where {T, S}?
What is the interpretation of the following two executions?
```julia
julia> findfirst(==(-7), 1:-1:-10)

julia> findfirst(==(-7), collect(1:-1:-10))
9

```

### PR DESCRIPTION
**Pull request by laborg**
Fixes #33808 
`findfirst` assumed increasing step size for `StepRange`.

### CODE REVIEW

**Review by mbauman — APPROVED** [review]

**Review by rfourquet — APPROVED** [review]

**Review by raminammour — COMMENTED** [review]
Maybe add:
```
isempty(r) && return nothing
```
Fixes things like:
````
julia> findfirst(==(2),1:-1:2)
0
````

File: test/ranges.jl [hunk 1]
```diff
@@ -305,6 +305,7 @@ end
         @test findfirst(==(7), 1:2:10) == 4
         @test findfirst(==(10), 1:2:10) == nothing
         @test findfirst(==(11), 1:2:10) == nothing
+        @test findfirst(==(-7), 1:-1:-10) == 9
```
Comment (dkarrasch) [inline test/ranges.jl]: maybe add a test where the range is empty, now that this is specifically addressed in the code?

**Review by dkarrasch — APPROVED** [review]

**Comment by mbauman** [comment]
Man, what's going on with CI?  6 ❌s, all different and seemingly unrelated!?  Let's try that again.

**Review by raminammour — APPROVED** [review]
